module/zfs/dmu_objset.c

```diff
@@ -616,7 +616,13 @@ dmu_objset_open_impl(spa_t *spa, dsl_dataset_t *ds, blkptr_t *bp,
 	mutex_init(&os->os_userused_lock, NULL, MUTEX_DEFAULT, NULL);
 	mutex_init(&os->os_obj_lock, NULL, MUTEX_DEFAULT, NULL);
 	mutex_init(&os->os_user_ptr_lock, NULL, MUTEX_DEFAULT, NULL);
-	os->os_obj_next_percpu_len = boot_ncpus;
+	/*
+	 * We allocate an array that is large enough to have a separate entry
+	 * for each cpu the system might eventually have, up to 32 cpus. This
+	 * allows us to provide good performance after cpus are hot-added
+	 * without using boundless amouunts of memory.
+	 */
+	os->os_obj_next_percpu_len = MAX(boot_ncpus, MIN(32, max_ncpus));
```

Comment:
I noticed that we are using this calculation here, but not in similar per-CPU data structures like multilists and aggsum's (which both use boot_ncpus).  I think that those general-purpose data structures are typically more important to performance than `os_obj_next_percpu`.  So should we use a similar calculation there?